src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/FileServerHandler.java

```diff
@@ -265,23 +269,142 @@ private void serveDefaultFavIcon(HttpExchange exchange, boolean writeBody)
         }
     }
 
-
     private void serveFile(HttpExchange exchange, Path path, boolean writeBody)
         throws IOException
     {
         var respHdrs = exchange.getResponseHeaders();
         respHdrs.set("Content-Type", mediaType(path.toString()));
         respHdrs.set("Last-Modified", getLastModified(path));
-        if (writeBody) {
-            exchange.sendResponseHeaders(200, Files.size(path));
-            try (InputStream fis = Files.newInputStream(path);
-                 OutputStream os = exchange.getResponseBody()) {
-                fis.transferTo(os);
-            }
-        } else {
+        respHdrs.set("Accept-Ranges", "bytes");
+        if (!writeBody) {
             respHdrs.set("Content-Length", Long.toString(Files.size(path)));
             exchange.sendResponseHeaders(200, -1);
+            return;
+        }
+        String rangeHeader = exchange.getRequestHeaders().getFirst("Range");
+        if (rangeHeader != null && handleRangedRequest(exchange, path, rangeHeader)) {
+            return;
+        }
+        exchange.sendResponseHeaders(200, Files.size(path));
+        try (InputStream fis = Files.newInputStream(path);
+             OutputStream os = exchange.getResponseBody()) {
+            fis.transferTo(os);
+        }
+    }
+
+    private boolean handleRangedRequest(HttpExchange exchange, Path path, String rangeHeader)
+            throws IOException
+    {
+        var reqHdrs = exchange.getRequestHeaders();
+        String ifRange = reqHdrs.getFirst("If-Range");
+        if (!(ifRange == null || ifRange.equals(getLastModified(path)))) {
+            return false; // last-modified does not match with the resource, send the entire file
+        }
+        long fileSize = Files.size(path);
+        List<RangeEntry> ranges = parseRangeHeader(rangeHeader, fileSize);
+        if (ranges == null) {
+            var respHdrs = exchange.getResponseHeaders();
+            respHdrs.set("Content-Range", "bytes */%d".formatted(fileSize));
+            exchange.sendResponseHeaders(416, -1);
+            return true;
+        }
+        servePartialContents(exchange, path, ranges);
+        return true;
+    }
+
+    private List<RangeEntry> parseRangeHeader(String rangeHeader, long fileSize) {
```

Comment:
It would be good to add a comment block before this method (no need for javadoc-style, // is enough) to give an example of all the valid values for the `rangeHeader` string. That would help in asserting that the method body does what it is supposed to do, and it will make easier to maintain should a bug be discovered later on.